### DISCUSSION TITLE
Domains vs launch menu entries

### DISCUSSION THREAD

**Discussion by TomiBelan** [wezterm/wezterm#2093] [Q&A] [answered]
Hi, nice terminal! But I'm confused by the distinction between domains and launch menu entries. Suppose I want these five tab types:

1. cmd.exe (default shell)
2. Powershell
3. MSYS2 (https://www.msys2.org/docs/terminals/)
4. ssh foo.example
5. ssh bar.example

As a user I don't think of them differently, but apparently "the wezterm way" is to configure 1 with default_prog, 2-3 with launch_menu and 4-5 with ssh_domains. My first attempt at configuring it was like this:

<details><summary>wezterm.lua</summary>

```lua
return {
  launch_menu = {
    {
      label = "Powershell",
      args = {"powershell"},
    },
    {
      label = "MSYS2",
      args = {"D:/msys64/msys2_shell.cmd", "-defterm", "-here", "-no-start", "-msys"},
      cwd = "D:/msys64/home/" .. os.getenv('USERNAME'),
    },
  },
  ssh_domains = {
    {
      name = "foo",
      remote_address = "foo.example",
      username = "example",
      multiplexing = "None",
    }
    {
      name = "bar",
      remote_address = "bar.example",
      username = "example",
      multiplexing = "None",
    }
  },
}
```

</details>

This works and shows all 5 entries in the launch menu. But there are two issues:

1. Ctrl+Shift+T is inconsistent. If I press it in a powershell or msys2 tab, I get a cmd.exe tab. If I press it in a ssh tab, I get another equivalent ssh tab. How do I make it always create the same kind as the current tab?

    - The default action of Ctrl+Shift+T is `SpawnTab="CurrentPaneDomain"`, so my first thought was "okay, maybe I need to create additional local domains for powershell and msys2 instead of using launch menu entries". But that doesn't seem to be possible. There is only one local domain.

2. If the current tab is ssh and I use the launcher menu to try to open powershell or msys2, wezterm tries to run powershell or msys2_shell.cmd on the ssh server. It seems I might be able to override this with `domain={DomainName="local"}` - is that true, and why is that necessary / not the default?

I'm getting the feeling that the wezterm implementation is leaking and doesn't match my mental model...

I'm interested both in workarounds to make my current config work, and thoughts on how wezterm's config model itself could change to be more intuitive. I posted this in discussions because I'm not really sure what I want, so it's not quite a complete bug report or feature request.

**Comment by wez** [comment, marked as answer]
> Ctrl+Shift+T is inconsistent. If I press it in a powershell or msys2 tab, I get a cmd.exe tab. If I press it in a ssh tab, I get another equivalent ssh tab. How do I make it always create the same kind as the current tab?

The `SpawnTab` assignment that is bound to the new tab button and `CTRL+SHIFT+T` is defined as running the `default_prog` for the domain associated with the current pane.

wezterm currently has no memory of what program was actually spawned into a pane; that is something that will likely get added as part of #2080, and once we have it I suspect that we'll also add an action to duplicate the command from the current pane as a new tab/pane.  I'm not 100% sold that that should become the default operation, as the launch menu can be (and is) used to spawn things like editors or `top` and I think it would be surprising for CTRL-SHIFT-T (which I think of as "give me a new shell") to duplicate top or a config editor tab.

> If the current tab is ssh and I use the launcher menu to try to open powershell or msys2, wezterm tries to run powershell or msys2_shell.cmd on the ssh server. It seems I might be able to override this with domain={DomainName="local"} - is that true, and why is that necessary / not the default?

The default domain for a SpawnCommand is `CurrentPaneDomain` which explains the behavior you're seeing.
You can explicitly set `domain={DomainName="local"}` to force the item to run in the local domain.

Why isn't that the default?  This is something that's evolved a bit over time and I think it is worth rethinking this default.
Changing this has the potential to break someone's config when they upgrade, so it needs a little bit of thought before just going ahead with making that change.

**Reply by TomiBelan** [reply]
Thanks for answering! That makes sense. I hope some of this is at least useful as food for thought.

Do you think there is a way to get something like my desired Ctrl+Shift+T behavior in today's wezterm? Maybe with Lua?

I agree about Ctrl+Shif+T meaning "give me a new shell" - it's just that I have multiple shells. One possible idea could be to add a boolean setting to launch menu entries to distinguish "shell-like" and "app-like" entries.

**Reply by TomiBelan** [reply]
I've got it.

Checking the foreground process is an imperfect heuristic, but it works fine for idle shells.

```lua
local wezterm = require 'wezterm';

local launch_powershell = {
  label = "Powershell",
  args = {"powershell"},
  domain = {DomainName="local"},
}

local launch_msys2 = {
  label = "MSYS2",
  args = {"D:/msys64/msys2_shell.cmd", "-defterm", "-here", "-no-start", "-msys"},
  cwd = "D:/msys64/home/" .. os.getenv('USERNAME'),
  domain = {DomainName="local"},
}

local function ctrl_shift_t(window, pane)
  if wezterm.version > "20220408-101518-b908e2dd" then
    if pane:get_domain_name() == "local" then
      local exe = pane:get_foreground_process_name()
      if exe and exe:lower():match("powershell.exe$") then
        window:perform_action(wezterm.action{SpawnCommandInNewTab=launch_powershell}, pane)
        return
      end
      if exe and exe:lower():match("\\msys64\\") then
        window:perform_action(wezterm.action{SpawnCommandInNewTab=launch_msys2}, pane)
        return
      end
    end
  end
  window:perform_action(wezterm.action{SpawnTab="CurrentPaneDomain"}, pane)
end

return {
  -- ...
  launch_menu = {
    launch_powershell,
    launch_msys2,
  },
  ssh_domains = {
    {
      name = "example",
      remote_address = "example.com",
      username = "example",
      -- ...
    },
  },
  keys = {
    {key="t", mods="CTRL|SHIFT", action=wezterm.action_callback(ctrl_shift_t)},
    {key="t", mods="SUPER", action=wezterm.action_callback(ctrl_shift_t)},
    -- ...
  },
}
```

**Reply by pwsandoval** [reply]
@TomiBelan Is this still your best solution to 2023?  or did you find a better one, I'm in the same situation, I want to use CTRL T to open the current shell (powershell, gitbash, nushell, etc)

**Reply by TomiBelan** [reply]
Yes I still use this config. I'm not aware of any better solution. (though I'll note I haven't really kept up with wezterm changelogs)
